flang/include/flang/Lower/Allocatable.h

```diff
@@ -12,28 +12,34 @@
 
 namespace mlir {
 class Value;
+class Type;
 class Location;
 } // namespace mlir
 
+namespace llvm {
+template <typename T>
+class ArrayRef;
+}
+
 namespace Fortran::parser {
 struct AllocateStmt;
 struct DeallocateStmt;
 } // namespace Fortran::parser
 
 namespace Fortran::lower {
 class AbstractConverter;
-namespace pft {
-struct Variable;
-}
+class FirOpBuilder;
 
-/// Generate fir to initialize the box (descriptor) of an allocatable variable.
-/// Initialization of such box has to be done at the beginning of the variable
-/// lifetime.
-/// The memory address of the box to be initialized must be provided as an
-/// input.
-void genAllocatableInit(Fortran::lower::AbstractConverter &,
-                        const Fortran::lower::pft::Variable &,
-                        mlir::Value boxAddress);
```

Comment:
This function is replaced by `genUnallocatedBox` to get the initial descriptor for an allocable (the main difference is that genAllocatableInit was storing this value inside the descriptor, that part is delegated to the caller when needed. For global the initial value is not stored, but returned with fir.has_value.)

To your second question, my understanding is that allocatable are necessarily unallocated when created.
C808:
> An initialization shall not appear if object-name is [...] an allocatable variable

9.7.1.3 point 2:
> At the beginning of execution of a program, allocatable variables are unallocated.

For pointer however, an initial target may be provided (I am not sure it can be provided for saved pointers, but that should be OK for us to support this anyway).

Reply:
Correct.